### PR TITLE
AMMP-663: device-based readings and communication over MQTT

### PR DESCRIPTION
**Pull request by judruizc**
Hey @svet-b, here is the basic functionality for getting device-based readings up to the queues. Then the datapusher, depending on the endpoint -- api, mqtt, influx..is in charge of re-factoring the payload. I wrote some notes on the process:
- first: problem is that we were “losing” the association between devices and readings when we were writing to the queues.  Modifications at the datapusher side were too late. We needed to keep this association **up to the moment of writing** to the queues (non-v and volatile)

- first approach was to separate reading the devices themselves from the shape/construction of the readout (the thing that was being written to the queues)
	This was very complicated as the jobs that read the devices (line 153+ get_readings) append the results to a readout_q and the readout is created on the spot using that. Hard to de-couple

-  second approach was to **modify the readout itself**, so it still contained the device data. Then the datapusher is in charge of reading it, and modifying it to best suit needs

- at the moment is only working for API endpoint, mqtt next

### CODE REVIEW

**Review by svet-b — COMMENTED** [review]
Thanks @judruizc ! Overall this looks like the right direction to me. In particular, I agree with changing the structure of the `readout` payload, and saving the full device-based readings in the new structure (while no longer attempting to make it "Influx-compatible).

Most of my comments are of a stylistic nature. I haven't dug into much detail on the changes in logic. To this end, could you share (maybe via another channel) an example `readout` that's generated in this way?

**Comment by judruizc** [comment]
> Thanks @judruizc ! Overall this looks like the right direction to me. In particular, I agree with changing the structure of the `readout` payload, and saving the full device-based readings in the new structure (while no longer attempting to make it "Influx-compatible).
> 
> Most of my comments are of a stylistic nature. I haven't dug into much detail on the changes in logic. To this end, could you share (maybe via another channel) an example `readout` that's generated in this way?

A "Raw readout" how I call it, looks like this at the moment. It's not ideal because as you can see, it may contain several entries (i.e. dicts) for a single Device. This is specially true for the _logger_ (still unknown when reading many devices), because the _raw readout_ is constructed as the code goes: Once to get the Snap version, another to get the cpu/memory/etc.. and a last one to get the reading_duration. (I assumed the Snap reading and the reading_duration to be under _logger_.)
`{
  "time": "2020-09-01T09:14:00Z",
  "device_readings": [
    {
      "dev_id": "logger",
      "comms_lggr_snap_rev": 0
    },
    {
      "dev_id": "logger",
      "comms_lggr_boot_time": 1594463915,
      "comms_lggr_cpu_load": 0.39,
      "comms_lggr_mem_usage": 77
    },
    {
      "dev_id": "logger_strato",
      "comms_lggr_cpu_temp": 57.458,
      "comms_lggr_disk_usage": 62.2
    },
    {
      "dev_id": "logger",
      "reading_duration": 0.042691
    }
  ],
  "meta": {}
}`

File: Pipfile [hunk 1]
```diff
@@ -30,4 +30,4 @@ ammp-edge = {path = "./src"}
 paho-mqtt = "*"
 
 [requires]
-python_version = "3.6"
+python_version = "3.8"
```
[the remark this answers is not in the document]
Comment (judruizc) [reply Pipfile]: The reason was that I only had 3.8 on my laptop's venvs.. so I wanted it to stop complaining, but of-course I meant to set it back (I already did, and installed a new venv with 3.6 locally)

File: src/ammp_edge.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 
 import node_mgmt
 from data_mgmt import DataPusher, NonVolatileQProc
-from reader import get_readings
+from reader.get_readings import get_readout
```
[the remark this answers is not in the document]
Comment (judruizc) [reply src/ammp_edge.py]: Should be fixed

**Comment by judruizc** [comment]
Hey Svet, an update on this. So we get now the device-based and the older API format readouts:

Device-based
```
{
  "time": "2020-10-06T10:46:00Z",
  "snap_rev": 0,
  "reading_duration": 0.049848,
  "meta": {
    "config_id": "7e2fdcf"
  },
  "device_readings": [
    {
      "dev_id": "logger",
      "boot_time": 1599053217,
      "cpu_load": 0.45,
      "memory_usage": 73
    },
    {
      "dev_id": "logger_strato",
      "cpu_temp": 53.692,
      "disk_usage": 67.1
    }
  ]
}```

API older format
```
{
  "time": "2020-10-06T10:46:00Z",
  "meta": {
    "config_id": "7e2fdcf"
  },
  "fields": {
    "comms_lggr_boot_time": 1599053217,
    "comms_lggr_cpu_load": 0.45,
    "comms_lggr_cpu_temp": 53.692,
    "comms_lggr_disk_usage": 67.1,
    "comms_lggr_mem_usage": 73,
    "snap_rev": 0,
    "reading_duration": 0.049848,
    "reading_offset": 0
  }
}```

I wasn't sure what to do with the `snap_rev`, `reading_duration` and `reading_offset` in the device-based one.

**Review by svet-b — COMMENTED** [review]
Thanks Juan, these look good!

> I wasn't sure what to do with the `snap_rev`, `reading_duration` and `reading_offset` in the device-based one.
I think having them at the top level like that is fine. We will have a handler function within the MQTT listener that can decide what (if anything) to do with them.

File: src/data_mgmt/helpers/convert_to_api_payload.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+		fields.update(rdg)
+	# get the old reading names for backwards compatibility
+	for key in fields:
+		for rdg in readings_from_config:
+			logger.debug(f"CONVERT TO API PAYLOAD. KEY: {key}, RDG: {readings_from_config[rdg]['var']}")
+			if readings_from_config[rdg]['var'] == key:
+				key = rdg
+	readout['fields'] = fields
+	# move snap_rev, reading_duration , and reading_offset under fields
+	readout['fields'].update({"snap_rev": readout['snap_rev']})
```
[the remark this answers is not in the document]
Comment (judruizc) [reply src/data_mgmt/helpers/convert_to_api_payload.py]: Thanks for the good tip. Adapted it.

File: src/reader/get_readings.py [hunk 4]
```diff
@@ -147,12 +147,14 @@ def get_readings(node):
...
             fields = readout_q.get(block=False)
-            readout['fields'].update(fields)
+            readout['device_readings'].append(fields)
         except queue.Empty:
             logger.warning('Not all devices returned readings')
 
-    readout['fields']['reading_duration'] = \
-        (arrow.utcnow() - arrow.get(readout['time'])).total_seconds()
+    # Augment payload with current config ID
+    readout['meta'].update({'config_id': node.config.get('config_id', '')})
```
[the remark this answers is not in the document]
Comment (judruizc) [reply src/reader/get_readings.py]: Also adapted

**Comment by judruizc** [comment]
Hey Svet, so this is a sum-up of what is attempted here. 

1.  We needed to change the way readings were save to the internal queue. Before it saved it in the almost flat dict with all measurements taken under fields. That way we lost the **devices** info (i.e. who measured what). 
So what I did was mostly change that, to save it in a new _device-based readout format_. This is mostly done in the `get_readings.py` module. 
The new format looks like (`reading_offset `is added at the end, once is about to be pushed): 
```
{
  "time": "2020-10-06T10:46:00Z",
  "snap_rev": 0,
  "reading_duration": 0.049848,
  "meta": {
    "config_id": "7e2fdcf"
  },
  "device_readings": [
    {
      "dev_id": "logger",
      "boot_time": 1599053217,
      "cpu_load": 0.45,
      "memory_usage": 73
    },
    {
      "dev_id": "logger_strato",
      "cpu_temp": 53.692,
      "disk_usage": 67.1
    }
  ]
}
```
2. We needed to keep consistency API payload. For this, if the `datapusher.py` module detects we want to push to API, it uses the new module `convert_to_api_payload.py` that basically transforms the readout into the familiar way. 

3. The addition under `network_host_finder.py` is just to fix the 00:00:00:00 MAC issue.

**Review by svet-b — CHANGES_REQUESTED** [review]

File: src/data_mgmt/datapusher.py [hunk 5]
```diff
@@ -172,7 +172,7 @@ def __push_readout(self, readout_to_push):
 
         elif self._dep.get('type') == 'mqtt':
             logger.debug(f"PUSH [MQTT]. Device-based readout: {readout_to_push}")
-            pub = self._mqtt_session.publish(f"a/{self._node.node_id}/data", json.dumps(readout_to_push), qos=1)
+            pub = self._mqtt_session.publish(f"a/{self._node.node_id}/data", json.dumps(readout_to_push), 1, False)
```
[the remark this answers is not in the document]
Comment (judruizc) [reply src/data_mgmt/datapusher.py]: Didn't know about the extra spaces. I corrected to:
```
            MQTT_QOS = 1
            MQTT_RETAIN = False
            MQTT_PUB_SUCCESS = 0
            logger.debug(f"PUSH [MQTT]. Device-based readout: {readout_to_push}")
            pub = self._mqtt_session.publish(f"a/{self._node.node_id}/data",
                                             json.dumps(readout, separators=(',', ':')),
                                             qos=MQTT_QOS, retain=MQTT_RETAIN)
```

File: src/reader/helpers/network_host_finder.py [hunk 6]
```diff
@@ -27,7 +27,10 @@ def arp_get_mac_from_ip(ip: str) -> str:
                 try:
                     this_ip, _, _, this_mac, _, _ = l.split()
                 except ValueError:
-                    logger.warn(f"Malformed ARP table entry: {l}. Skipping")
+                    logger.warning(f"Malformed ARP table entry: {l}. Skipping")
+                    continue
+                if this_mac == '00:00:00:00:00:00':
```
[the remark this answers is not in the document]
Comment (judruizc) [reply src/reader/helpers/network_host_finder.py:32]: Makes much sense. Should be corrected now

File: src/data_mgmt/datapusher.py [hunk 7]
```diff
@@ -5,31 +5,46 @@
...
         elif dep.get('type') == 'influxdb':
             self._session = InfluxDBClient(**dep['client_config'])
+        elif dep.get('type') == 'mqtt':
+            if os.environ.get('MQTT_LEVEL'):
+                try:
+                    mqtt_cert_path = os.path.join(os.getenv('SNAP_COMMON', './'), 'resources', 'ca-' + os.environ.get('MQTT_LEVEL') + '.crt')
+                    logger.debug(f"MQTT level: {os.environ.get('MQTT_LEVEL')}")
+                except Exception:
+                    logger.warning(f"Failed to set mqtt level to {os.environ['MQTT_LEVEL']}", exc_info=True)
+            self._mqtt_session = mqtt.Client(client_id="ammp", clean_session=False, transport="tcp")
```
[the remark this answers is not in the document]
Comment (judruizc) [reply src/data_mgmt/datapusher.py]: Yeah I was a bit confused by this client_id. I'm using now the node id for it as you suggest:
`self._mqtt_session = mqtt.Client(client_id=self._node.node_id, clean_session=False, transport="tcp")`

**Comment by judruizc** [comment]
Hey @svet-b , thanks for the feedback -- always giving me the "why" :) 

I hope most of these issues are corrected now. Please have a look when you can.

**Pull request merged**